### ISSUE TITLE
您好。当我修改beam_size=2的时候，出现了错误。

### ISSUE THREAD

**Issue by enjoy4** [ruotianluo/self-critical.pytorch#74]
File "train.py", line 224, in <module>
    train(opt)
  File "train.py", line 168, in train
    val_loss, predictions, lang_stats = eval_utils.eval_split(dp_model, crit, loader, eval_kwargs)
  File "/media/enjoy/32F872CFF87290B9/CODE/self-critical-beam.pytorch-master/eval_utils.py", line 107, in eval_split
    print('\n'.join([utils.decode_sequence(loader.get_vocab(), _['seq'].unsqueeze(0))[0] for _ in model.done_beams[i]]))
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.py", line 532, in __getattr__
    type(self).__name__, name))
AttributeError: 'DataParallel' object has no attribute 'done_beams'
Terminating BlobFetcher
希望帮忙解答，感谢！

**Comment by jamiechoi1995** [comment]
me too. 你是在reinforcement learning阶段对吗

**Comment by ruotianluo** [comment]
你可以把eval 的输入从dpmodel改成model

**Comment by jamiechoi1995** [comment]
@ruotianluo 
after modifying dp_model to model, I got CUDA out of memory error,
guess this is caused by putting all the data in one GPU?

can I set the batch_size = 1 in eval_utils? https://github.com/ruotianluo/self-critical.pytorch/blob/a99a7603ba76d0b051b0b0a8a141fc27c85062a9/eval_utils.py#L71

would it affect the data loading?
Thanks!

**Comment by ruotianluo** [comment]
Try removing done beams related stuff in eval_utils. If that doesn't work. I'm not sure if beam search works for multigpu, but give it a try. 

Ruotian Luo

**Comment by jamiechoi1995** [comment]
For your informantion:

when using multigpu with beam size > 1
I got the following error:

``  File "train.py", line 246, in <module>
    train(opt)
  File "train.py", line 179, in train
    val_loss, predictions, lang_stats = eval_utils.eval_split(dp_model, crit, loader, eval_kwargs)
  File "/jamiecai/feat/test/eval_utils.py", line 102, in eval_split
    seq = model(fc_feats, att_feats, att_masks, opt=eval_kwargs, mode='sample')[0].data
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.py", line 477, in __call__
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/data_parallel.py", line 124, in forward
    return self.gather(outputs, self.output_device)
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/data_parallel.py", line 136, in gather
    return gather(outputs, output_device, dim=self.dim)
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/scatter_gather.py", line 67, in gather
    return gather_map(outputs)
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/scatter_gather.py", line 62, in gather_map
    return type(out)(map(gather_map, zip(*outputs)))
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/scatter_gather.py", line 54, in gather_map
    return Gather.apply(target_device, dim, *outputs)
  File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/_functions.py", line 52, in forward
    assert all(map(lambda i: i.is_cuda, inputs))``

when I replace dp_model with model, problem solved but got the out of memory issue mention above.

**Comment by jamiechoi1995** [comment]
> 
> Try removing done beams related stuff in eval_utils. If that doesn't work. I'm not sure if beam search works for multigpu, but give it a try. Ruotian Luo
> […](#)
> On Mar 25, 2019, at 10:41 PM, Jamie ***@***.***> wrote: @ruotianluo after modifying dp_model to model, I got CUDA out of memory error, guess this is caused by putting all the data in one GPU? can I set the batch_size = 1 in eval_utils? https://github.com/ruotianluo/self-critical.pytorch/blob/a99a7603ba76d0b051b0b0a8a141fc27c85062a9/eval_utils.py#L71 would it affect the data loading? Thanks! — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

Thank you for your reply.
However, if I remove the beams related stuff, it will affect the performance of reinforcement training, right?

**Comment by ruotianluo** [comment]
Beam search is not used during reinforcement learning in my implementation actually.  It's only used during evaluation

**Comment by jamiechoi1995** [comment]
I solve it by adding ``eval_kwargs['beam_size']=1`` after 
https://github.com/ruotianluo/self-critical.pytorch/blob/a99a7603ba76d0b051b0b0a8a141fc27c85062a9/train.py#L167

**Comment by ruotianluo** [comment]
I thought you want beam size greater than 1. In fact I think during training, beam size is set to one in default. Also using --beam_size 1 with train.py should work.

**Comment by congve1** [comment]
maybe decrease the batch size will solve the out of memory issue?

**Comment by jamiechoi1995** [comment]
> 
> 
> I thought you want beam size greater than 1. In fact I think during training, beam size is set to one in default. Also using --beam_size 1 with train.py should work.

Yes, I want beam size > 1 to see if there are any boost in performance

**Comment by jamiechoi1995** [comment]
> 
> 
> maybe decrease the batch size will solve the out of memory issue?

In this way, I have to significantly decrease the batch size to make it work.

**Comment by enjoy4** [comment]
thx.  actually  I find  the attribute 'done_beam' was in model.module.  when I replace the code by( for _ in model.module.done_beams) it works

**Comment by congve1** [comment]
@jamiechoi1995 
add `"verbose_beam":0`  in the `eval_kwargs`, this will not print the `done_beams` which may solve this issue.
If you do want to print the `done_beams`,  try to pass the `dp_model.module` instead of `dp_model` when calling the `eval_split` method.
If these do not solve the problem, I have no idea to do with the problem.

**Comment by enjoy4** [comment]
@ruotianluo thanks for your code！  I have learned a lot through all the code you wrote！

**Comment by jamiechoi1995** [comment]
@congve1 adding ``"verbose_beam":0`` in the ``eval_kwargs`` does not work, it raise this error -> https://github.com/ruotianluo/self-critical.pytorch/issues/74#issuecomment-476472695

**Comment by jamiechoi1995** [comment]
> 
> 
> For your informantion:
> 
> when using multigpu with beam size > 1
> I got the following error:
> 
> ` File "train.py", line 246, in <module> train(opt) File "train.py", line 179, in train val_loss, predictions, lang_stats = eval_utils.eval_split(dp_model, crit, loader, eval_kwargs) File "/jamiecai/feat/test/eval_utils.py", line 102, in eval_split seq = model(fc_feats, att_feats, att_masks, opt=eval_kwargs, mode='sample')[0].data File "/usr/local/lib/python2.7/dist-packages/torch/nn/modules/module.py", line 477, in __call__ result = self.forward(*input, **kwargs) File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/data_parallel.py", line 124, in forward return self.gather(outputs, self.output_device) File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/data_parallel.py", line 136, in gather return gather(outputs, output_device, dim=self.dim) File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/scatter_gather.py", line 67, in gather return gather_map(outputs) File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/scatter_gather.py", line 62, in gather_map return type(out)(map(gather_map, zip(*outputs))) File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/scatter_gather.py", line 54, in gather_map return Gather.apply(target_device, dim, *outputs) File "/usr/local/lib/python2.7/dist-packages/torch/nn/parallel/_functions.py", line 52, in forward assert all(map(lambda i: i.is_cuda, inputs))`
> 
> when I replace dp_model with model, problem solved but got the out of memory issue mention above.

I think the above error `assert all(map(lambda i: i.is_cuda, inputs))` is caused by similar reason as https://github.com/NVIDIA/flownet2-pytorch/issues/146
